core/include/vecmem/memory/details/unique_alloc_deleter.hpp

```diff
@@ -10,12 +10,116 @@
+ * @tparam T The type to deallocate.
+ *
+ * @note This deleter class supports non-array types, array types of unknown
+ * bound, but not array types of known bound.
+ *
+ * @note This deleter class only supports types that are both trivially
+ * constructible and trivially destructible.
+ */
 template <typename T>
-using unique_alloc_deleter = std::enable_if_t<
-    std::is_trivially_constructible_v<std::remove_extent_t<T>> &&
-        std::is_trivially_destructible_v<std::remove_extent_t<T>>,
-    unique_alloc_deleter_impl>;
-}
+struct unique_alloc_deleter {
+    /*
+     * Similarly, we cannot destroy objects, so we need to ensure that
+     * deallocation of objects is semantically equivalent to their destruction.
+     */
+    static_assert(std::is_trivially_destructible_v<std::remove_extent_t<T>>,
+                  "Allocation pointer type must be trivially destructible.");
+
+public:
+    /**
+     * @brief Default-construct a new unique allocation deleter.
+     *
+     * This constructor should only really be used in cases where we are
+     * nulling a unique pointer.
+     */
+    unique_alloc_deleter(void) = default;
+
+    /**
+     * @brief Construct a new unique allocation deleter.
+     *
+     * This is the standard constructor for this class, which stores all the
+     * necessary data to deallocate an allocation.
+     *
+     * @param mr The memory resource to use for deallocation.
+     * @param p The pointer to deallocate.
+     * @param s The size of the allocation.
+     * @param a The alignment of the allocation.
+     */
+    unique_alloc_deleter(memory_resource& mr, std::size_t s, std::size_t a = 0)
+        : m_mr(&mr), m_size(s), m_align(a) {}
+
+    /**
+     * @brief Copy a unique allocation deleter.
+     *
+     * @param i The allocation deleter to copy.
+     */
+    unique_alloc_deleter(const unique_alloc_deleter& i) = default;
+
+    /**
+     * @brief Move a unique allocation deleter.
+     *
+     * @param i The allocation deleter to move.
+     */
+    unique_alloc_deleter(unique_alloc_deleter&& i) = default;
+
+    /**
+     * @brief Copy-assign a unique allocation deleter.
+     *
+     * @param i The object to copy into the current one.
+     *
+     * @return A reference to this object.
+     */
+    unique_alloc_deleter& operator=(const unique_alloc_deleter& i) = default;
+
+    /**
+     * @brief Move-assign a unique allocation deleter.
+     *
+     * @param i The object to move into the current one.
+     *
+     * @return A reference to this object.
+     */
+    unique_alloc_deleter& operator=(unique_alloc_deleter&& i) = default;
+
+    /**
+     * @brief Activate the deletion mechanism of the deleter.
+     *
+     * @param p The pointer to deallocate.
+     */
+    void operator()(void* p) const {
+        assert(m_mr != nullptr);
+
+        /*
+         * As before, if this happens... Something has gone VERY wrong.
+         */
+        if (m_mr == nullptr) {
+            return;
+        }
```

Comment:
Hmm, I am not sure that's a great idea - in general I don't think a library like vecmem should really be terminating the program that is using it, that seems like dangerous behaviour. :thinking:

Reply:
But this is a silent failure, with possible memory corruption?

Reply:
I don't _think_ this could ever lead to memory corruption. If I am reasoning about this correctly this only really fires if someone has already done something really stupid (something that violate the unique pointer's semantics) and invalidated the state of the pointer. So memory would already be lost, for example.

Not saying this is a great solution, but I am thinking this could be a case of "if you've come this far you've already fucked up beyond what is reasonable, so whatever happens is undefined behaviour".

Reply:
Ok.